src/agent/src/watcher.rs

```diff
@@ -0,0 +1,487 @@
+    #[async_recursion]
+    async fn add_path(&mut self, path: &Path) -> Result<()> {
+        if path.is_file() {
+            self.add_file(path)
+                .await
+                .with_context(|| format!("Failed to add file {}", path.display()))?;
+        } else {
+            let mut entries = fs::read_dir(path)
+                .await
+                .with_context(|| format!("Failed to read dir: {}", path.display()))?;
+
+            while let Some(entry) = entries.next_entry().await? {
+                self.add_path(entry.path().as_path()).await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Adds a single file path entry
+    async fn add_file(&mut self, file: &Path) -> Result<()> {
+        ensure!(
+            self.files.len() <= MAX_ENTRIES_PER_STORAGE,
+            "Too many file system entries to watch (must be < {})",
+            MAX_ENTRIES_PER_STORAGE
+        );
+
+        let entry = File {
+            file: file.to_path_buf(),
+            // Set minimum value to perform initial copy during next check
+            modified: SystemTime::UNIX_EPOCH,
+        };
+
+        self.files.push(entry);
+        Ok(())
+    }
+
+    async fn check(&mut self) -> Result<usize> {
+        let mut count = 0;
+        for i in 0..self.files.len() {
+            if self.check_file(i).await? {
+                count += 1;
+            }
+        }
+
+        Ok(count)
+    }
+
+    async fn check_file(&mut self, i: usize) -> Result<bool> {
+        let entry = &mut self.files[i];
+
+        let current = modified(&entry.file).await?;
+        if current <= entry.modified {
+            return Ok(false);
+        }
+
+        // Update entry
+        entry.modified = current;
+
+        // We no longer need mutable borrow, reborrow.
+        let entry = &self.files[i];
+
+        // File to file copy
+        if self.source.is_file() {
+            fs::copy(&self.source, &self.mount_point).await?;
+            return Ok(true);
+        }
+
+        // Dir to dir copy
+        let source_file_path = entry.file.as_path();
+        let dest_file_path = self.make_dest_path(&source_file_path)?;
+
+        if let Some(path) = dest_file_path.parent() {
+            fs::create_dir_all(path).await?;
+        }
+
+        fs::copy(source_file_path, dest_file_path).await?;
+        Ok(true)
+    }
+
+    fn make_dest_path(&self, source_file_path: impl AsRef<Path>) -> Result<PathBuf> {
+        let relative_path = source_file_path
+            .as_ref()
+            .strip_prefix(&self.source)
+            .with_context(|| {
+                format!(
+                    "Failed to get prefix: {} - {}",
+                    source_file_path.as_ref().display().to_string(),
+                    &self.source.display()
+                )
+            })?;
+        let dest_file_path = Path::new(&self.mount_point).join(relative_path);
+
+        Ok(dest_file_path)
+    }
+}
+
+#[derive(Default, Debug)]
+struct Entries(Vec<Entry>);
```

Comment:
An entry is created based on a storage objects source and destination, provided from the kata runtime. Based on this, these will always be unique within the sandbox, since the mount is renamed as $containerId-$uniqueID-$containerMountpoint.